dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/RelationshipRestRepository.java

```diff
@@ -60,4 +94,100 @@ public RelationshipRest findOne(Context context, Integer integer) {
+
+        HttpServletRequest req = getRequestService().getCurrentRequest().getHttpServletRequest();
+        ObjectMapper mapper = new ObjectMapper();
+        RelationshipRest relationshipRest = null;
+        try {
+            relationshipRest = mapper.readValue(req.getInputStream(), RelationshipRest.class);
+        } catch (IOException e1) {
+            throw new UnprocessableEntityException("error parsing the body");
+        }
+
+        Relationship relationship = new Relationship();
+        Item leftItem = itemService.find(context, UUIDUtils.fromString(req.getParameter("leftItem")));
+        Item rightItem = itemService.find(context, UUIDUtils.fromString(req.getParameter("rightItem")));
+        RelationshipType relationshipType = relationshipTypeService
+            .find(context, Integer.parseInt(req.getParameter("relationshipType")));
+
+        EPerson ePerson = context.getCurrentUser();
+        if (authorizeService.authorizeActionBoolean(context, leftItem, Constants.WRITE) &&
+            authorizeService.authorizeActionBoolean(context, rightItem, Constants.WRITE)) {
+            relationship.setLeftItem(leftItem);
+            relationship.setRightItem(rightItem);
+            relationship.setRelationshipType(relationshipType);
+            relationship = relationshipService.create(context, relationship);
+            return relationshipConverter.fromModel(relationship);
+        } else {
+            throw new AccessDeniedException("You do not have write rights on this relationship's items");
+        }
+
+    }
+
+    @Override
+    protected RelationshipRest put(Context context, HttpServletRequest request, String apiCategory, String model,
+                                   Integer id,
+                                   JsonNode jsonNode)
+        throws RepositoryMethodNotImplementedException, SQLException, AuthorizeException {
+
+        HttpServletRequest req = getRequestService().getCurrentRequest().getHttpServletRequest();
+        ObjectMapper mapper = new ObjectMapper();
+        RelationshipRest relationshipRest = null;
+        try {
+            relationshipRest = mapper.readValue(jsonNode.toString(), RelationshipRest.class);
+        } catch (IOException e) {
+            throw new UnprocessableEntityException("error parsing the body");
+        }
+        Relationship relationship = relationshipService.find(context, id);
+        if (relationship == null) {
+            throw new ResourceNotFoundException(apiCategory + "." + model + " with id: " + id + " not found");
+        }
+        Item leftItem = itemService.find(context, UUIDUtils.fromString(req.getParameter("leftItem")));
+        Item rightItem = itemService.find(context, UUIDUtils.fromString(req.getParameter("rightItem")));
+        RelationshipType relationshipType = relationshipTypeService
+            .find(context, Integer.parseInt(req.getParameter("relationshipType")));
```

Comment:
This again looks like a `PUT` request with an empty body.  As discussed in the last DSpace 7 meeting, we need to avoid this type of design as it's not proper usage of `PUT`.  *Both* `PUT` and `POST` should receive their necessary creation/update data in the request body -- params should only be used if there's a reason why the request body won't work right.  In this situation, it seems like the `leftItem`, `rightItem` and `relationshipType` should be a part of the request body (unless I'm missing a reason why that won't work properly).

See again discussion notes at https://wiki.duraspace.org/display/DSPACE/2019-01-17+DSpace+7+Working+Group+Meeting

Reply:
@tdonohue I verified your new documentation for this request, and it's still unclear to me how we can specify the left and right item simultaneously while making sure it's clear which item is being created

If we would perform the call below, it's not clear which item should occur left or right:
`curl -i -X POST "http://localhost:8080/api/core/relationships" -H "Content-Type:text/uri-list" http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f \n http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067 \n http://localhost:8080/rest/api/core/relationshiptypes/1`

Can you please share how you would prefer to specify the items?

Reply:
Hi @benbosman, you are correct `text/uri-list` cannot be used in this scenario. It was my mistake in saying/implying that it could.

However, I think we *still need to avoid passing these values in as parameters* (as it would result in an empty request body).  

So, taking a step back, I'm starting to wonder if this POST is on the proper endpoint.  If we wanted a true `POST` to `/relationships`, I think it'd have to send some sort of JSON that looks like this:

```
{
  "relationshipType": "http://localhost:8080/rest/api/core/relationshiptypes/1",
  "leftItem": "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f",
  "rightItem": "http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067",
}
```

But, that seems a bit **ugly** to me.

So, I wonder if we should reconsider POST/PUT to the `/relationships` endpoint altogether, and instead create Relationships from the `/items` endpoint.  So, for example, we could create the same relationship above as a `POST` to the *leftItem*, something like this:

`curl -i -X POST "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f/relationships?type=1" -H "Content-Type:text/uri-list" -d "http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067"`

In this example we still do need *one* parameter, to specify the type of relationship to create. You'll notice we are creating the relationship from the leftItem to the rightItem.

All that said, I'd also be curious to hear what @abollini thinks here -- as I'm not entirely sure whether he's got a related example we could replicate.

Reply:
@tdonohue It also seems weird to me stating that the left item is always the endpoint while the right item is always in the uri-list.

We can perhaps go for an alternative:
`curl -i -X POST "http://localhost:8080/api/core/relationships?leftItem=<:itemuuid>&rightItem=<:itemuuid>" -H "Content-Type:text/uri-list" -d "http://localhost:8080/rest/api/core/relationshiptypes/1"`
It avoids an empty body and ensures it's clear what each part of the request is. Would that seem like a viable alternative to you?

Reply:
@benbosman : just trying to understand your comment about `POST`ing to the `leftItem` being "weird".  Is there any specific meaning associated with the "left" vs "right" item? I had thought they were just labels (with no extra meaning) to tell the two Items apart?  If it's just a label, then I don't know why `POST`ing one item to the other is a bad thing...as it'd essentially just be saying "I'm creating a new relationship between *this item* (the one you are POSTing to) and this other item (the URI you are passing to it)".

I'm not sure I like the idea of passing multiple fields on the querystring.  So, if we don't like the idea of posting one item to the other...then I think we should go back to my initial suggestion and simply POST everything as JSON (and leave out "text/uri-list" content type altogether).  To me, because this Relationship object is like a link (between objects), I thought text/uri-list might be a better path. But, since Relationship is also a seperate object/endpoint (with additional properties), maybe we should just treat all these fields as attributes/properties and send them in the JSON body.  So, I'm saying that if "text/uri-list" isn't going to work, maybe we need to just POST this:
```
{
  "relationshipType": "http://localhost:8080/rest/api/core/relationshiptypes/1",
  "leftItem": "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f",
  "rightItem": "http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067",
}
```

Reply:
@tdonohue Sorry for the vague term "weird" :)
I was just imagining a relationship type where the left item is a project and the right item is an author. When someone is submitting the project, they may need to post the new project as a relation to the existing author.
Or alternatively if the relationship type would be defined where the left item is an author and the right item is a project. When someone is submitting the author, they may need to post the new author as a relation to the existing project.
Creating a relationship between two items doesn't sound like a change to one of the items to me.
But I'm definitely not stating this is not an option for me. I can live with this since it works and solves the problem at hand.

If I'm not mistaken, the solution you proposed with the JSON body matches the second solution from 
https://github.com/DSpace/Rest7Contract/pull/33#discussion_r239505976
Neither Andrea nor I was able to identify a good implementation method for this. Also it would introduce a third format to identify the related object.

I was also wondering with your example
`curl -i -X POST "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f/relationships?type=1" -H "Content-Type:text/uri-list" -d "http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067"`
whether there's a particular reason to still specify the type as a parameter? I think this one would be the most straightforward to migrate to http://localhost:8080/rest/api/core/relationshiptypes/1

Reply:
@tdonohue I think the solution you mention would work:
`curl -i -X POST "http://localhost:8080/rest/api/core/relationshiptypes/1/relationships" -H "Content-Type:text/uri-list" -d "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f \n http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067"`
One question before writing it in stone: does this impact the method to update a relationship as well? For that method, it's possible to update only the left or only the right item, so that cannot be altered to use an ordered list at the moment.

I would actually still offer `GET /api/core/relationships/search/byName?relationshipname=<:relationshipname>`. This query can work directly on a name, and supports mixing relationships from 2 types with the same name (e.g. publication vs author and publication vs org unit both using isAuthorOfPublication).
At the moment we don't have a use case to require  `GET /api/core/relationshiptypes/[:id]/relationships`, but there's no problem in creating it as well

Reply:
@benbosman : Comparing Relationship endpoint to Item, Collection and Community, I've come to the realization that we can better align Relationship to act similar to those endpoints.  Here's my final proposal:

- `POST`: Creating a Relationship requires a valid RelationshipType ID, similar to how POST of Item, Collection and Community require a parent UUID: `curl -i -X POST "http://localhost:8080/rest/api/core/relationships?relationshiptype=1" -H "Content-Type:text/uri-list" -d "http://localhost:8080/rest/api/core/items/0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f \n http://localhost:8080/rest/api/core/items/e98b0f27-5c19-49a0-960d-eb6ad5287067"`
    - I've changed this POST request to pass the RelationshipType ID on the querystring, as that aligns better with how we POST Item, Collection & Community, as all of those objects also take in a querystring ID of the "parent" object.  Relationships seem to have a "parent" object of a RelationshipType (in that you cannot create a Relationship without first creating the RelationshipType).
    - If you prefer the POST on the `relationshiptype/[id]/relationship` endpoint, we could use that instead here.  The below PUT and DELETE would still be unchanged however.
- `PUT`: Updating a Relationship (of a given ID) is also similar to Item, Collection & Community: `PUT "http://localhost:8080/rest/api/core/relationships/[:id]"`  I'd assume that PUT however would likely need to take JSON data, unless the only thing you can update is the left/right item, in which case we could require this to be `text/uri-list` of the same format as `POST`
- `DELETE`: Again, similar to Item, Collection & Community: `DELETE "http://localhost:8080/rest/api/core/relationships/[:id]"`

As for your `/relationships/search/byName` endpoint...now that I understand it better, I have no objections to keeping it.

Reply:
Thanks @tdonohue 

- I agree with your solution for the `POST` with the relationshiptype as a parameter and the items in the uri-list body. This makes it similar to a POST of Item, Collection and Community as you mentioned
- For `PUT`:
  - We can drop the ability to change the relationshipType (it's currently supported, but it's probably not something we'll actually use in the UI). It's still possible to delete the relationship and create a new one if the relationshipType should change
  - We can only specify the items in the uri-list body if we make it mandatory to contain both items. Currently it's possible to only update e.g. the left item. So this seems like a viable solution as well

I will update the contract accordingly so it can be approved and we can start the development

Reply:
Sounds good @benbosman!  Thanks